src/backend/vulkan/mod.rs

```diff
@@ -532,6 +534,92 @@ impl PhysicalDevice {
         unsafe { instance.get_physical_device_properties2(self.handle(), props) }
     }
 
+    /// Get a list of supported formats.
+    ///
+    /// Checks a list of formats (currently hardcoded to be Argb8888 and Argb2101010 from DrmFourcc) to see if they are compatible with
+    /// the current device.
+    ///
+    /// Returns a vec of [`vk::Format`], with a length of 0 if none are found.
+    ///
+    /// ```
+    /// use smithay::{
+    ///     backend::vulkan::{version::Version, Instance, PhysicalDevice},
+    ///     reexports::ash::vk,
+    /// };
+    ///
+    /// let instance = Instance::new(Version::VERSION_1_3, None).unwrap();
+    ///
+    /// let physical_device = PhysicalDevice::enumerate(&instance)
+    ///     .unwrap()
+    ///     .next()
+    ///     .expect("No physical devices");
+    ///
+    /// let formats = physical_device.get_supported_formats();
+    ///
+    /// assert_eq!(
+    ///     formats,
+    ///     vec![
+    ///         vk::Format::B8G8R8A8_SRGB,
+    ///         vk::Format::A8B8G8R8_SRGB_PACK32,
+    ///         vk::Format::R8G8B8A8_SRGB,
+    ///         vk::Format::A2R10G10B10_UNORM_PACK32
+    ///     ]
+    /// );
+    /// ```
+    pub fn get_supported_formats(&self) -> Vec<vk::Format> {
+        let instance = self.instance().handle();
+        // TODO: maybe not use hardcoded formats?
+        let formats = vec![
+            DrmFourcc::Argb8888,
+            DrmFourcc::Bgra8888,
+            DrmFourcc::Rgba8888,
+            DrmFourcc::Xbgr8888,
+            DrmFourcc::Argb2101010,
+            DrmFourcc::Bgra1010102,
+            DrmFourcc::Rgba1010102,
+            DrmFourcc::Xrgb2101010,
+        ];
+        let mut supported = vec![];
+
+        for format in formats {
+            let vk_format = get_vk_format(format);
+
+            if vk_format.is_none() {
+                continue;
+            }
+
+            let format_info = vk::PhysicalDeviceImageFormatInfo2::builder()
+                .usage(
+                    vk::ImageUsageFlags::TRANSFER_SRC
+                        | vk::ImageUsageFlags::TRANSFER_DST
+                        | vk::ImageUsageFlags::SAMPLED,
+                )
+                .format(vk_format.expect("vk_format to exist"))
+                .ty(vk::ImageType::TYPE_2D);
+
+            let mut image_format_prop = vk::ImageFormatProperties2::default();
+            let res = unsafe {
+                instance.get_physical_device_image_format_properties2(
+                    self.handle(),
+                    &format_info,
+                    &mut image_format_prop,
+                )
+            };
+
+            match res {
+                Ok(_) => {
+                    if image_format_prop.image_format_properties.max_extent.depth == 1 {
+                        supported.push(vk_format.expect("vk_format to exist"));
+                    };
+                }
+                Err(_) => {}
+            };
+        }
+
+        supported.dedup();
```

Comment:
Let's explain why we dedup: Most DrmFourcc formats have an opaque and alpha variant. Vulkan does not have alpha and opaque variants of formats and instead represents opaque formats by swizzling the alpha channel.